packages/itwinui-react/src/core/Tabs/Tabs.tsx

```diff
@@ -177,46 +86,147 @@ export const Tabs = (props: TabsProps) => {
     ...rest
   } = props;
 
-  const isClient = useIsClient();
+  const {
+    orientation,
+    type,
+    activeValue: activeValueContext,
+    setActiveValue: setActiveValueContext,
+    overflowOptions,
+    setStripeProperties,
+  } = useSafeContext(TabsContext);
+
+  const items = React.useMemo(
+    () => (Array.isArray(children) ? children : [children]),
+    [children],
+  );
 
-  const tablistRef = React.useRef<HTMLUListElement>(null);
+  const activeValue = activeValueProp || activeValueContext;
+  const setActiveValue = setActiveValueProp || setActiveValueContext;
+
+  const isClient = useIsClient();
+  const tablistRef = React.useRef<HTMLDivElement>(null);
   const [tablistSizeRef, tabsWidth] = useContainerWidth(type !== 'default');
-  const refs = useMergedRefs(tablistRef, tablistSizeRef);
+  const refs = useMergedRefs(ref, tablistRef, tablistSizeRef);
+  const newActiveIndex = React.useRef(0);
+
+  const [focusedValue, setFocusedValue] = React.useState<string | undefined>();
+  const newFocusedIndex = React.useRef<number | undefined>();
+  const [hasSublabel, setHasSublabel] = React.useState(false); // used for setting size
 
-  const [currentActiveIndex, setCurrentActiveIndex] = React.useState(() =>
-    activeIndex != null
-      ? getBoundedValue(activeIndex, 0, labels.length - 1)
-      : 0,
-  );
   useIsomorphicLayoutEffect(() => {
-    if (activeIndex != null && currentActiveIndex !== activeIndex) {
-      setCurrentActiveIndex(getBoundedValue(activeIndex, 0, labels.length - 1));
+    // Should only call on the first render since activeValue cannot be initialized in TabsWrapper
+    const activeItem = items.find(
+      (item) => React.isValidElement(item) && item.props.isActive,
+    );
+
+    if (React.isValidElement(activeItem) && !!activeItem) {
+      const value = activeItem.props.value as string;
+      if (value !== activeValue) {
+        setActiveValue(value);
+      }
+    } else {
+      React.isValidElement(items[0]) &&
+        setActiveValue(items[0].props.value as string);
     }
-  }, [activeIndex, currentActiveIndex, labels.length]);
+  }, []);
+
+  // When activeValue change
+  useIsomorphicLayoutEffect(() => {
+    items.forEach((item, index) => {
+      if (React.isValidElement(item) && item.props.value === activeValue) {
+        newActiveIndex.current = index;
+        return;
+      }
+    });
+  }, [items, activeValue]);
+
+  // When focusedValue change
+  useIsomorphicLayoutEffect(() => {
+    items.forEach((item, index) => {
+      if (React.isValidElement(item) && item.props.value === focusedValue) {
+        newFocusedIndex.current = index;
+        return;
+      }
+    });
+  }, [items, focusedValue]);
 
   // CSS custom properties to place the active stripe
-  const [stripeProperties, setStripeProperties] = React.useState({});
   useIsomorphicLayoutEffect(() => {
-    if (type !== 'default' && tablistRef.current != undefined) {
+    console.log(activeValue);
```

Comment:
nit: Creating a comment so we don't forget to delete it before merging

Reply:
Removed